src/Workspaces/Core/Portable/EmbeddedLanguages/RegularExpressions/RegexParser.cs

```diff
@@ -0,0 +1,1927 @@
+                        return new RegexSimpleEscapeNode(
+                            backslashToken, ConsumeCurrentToken(allowTrivia: false).With(kind: RegexKind.TextToken));
+
+                    default:
+                        // trivia is not allowed anywhere in a character class
+                        _lexer.Position--;
+                        return ScanCharEscape(backslashToken, allowTriviaAfterEnd: false);
+                }
+            }
+
+            if (!afterRangeMinus &&
+                !isFirst &&
+                _currentToken.Kind == RegexKind.MinusToken &&
+                _lexer.IsAt("["))
+            {
+                // have a trailing subtraction.
+                // trivia is not allowed anywhere in a character class
+                return ParseCharacterClassSubtractionNode(
+                    ConsumeCurrentToken(allowTrivia: false));
+            }
+
+            // From the .net regex code:
+            // This is code for Posix style properties - [:Ll:] or [:IsTibetan:].
+            // It currently doesn't do anything other than skip the whole thing!
+            if (!afterRangeMinus && _currentToken.Kind == RegexKind.OpenBracketToken && _lexer.IsAt(":"))
+            {
+                var beforeBracketPos = _lexer.Position - 1;
+                // trivia is not allowed anywhere in a character class
+                ConsumeCurrentToken(allowTrivia: false);
+
+                var captureName = _lexer.TryScanCaptureName();
+                if (captureName.HasValue && _lexer.IsAt(":]"))
+                {
+                    _lexer.Position += 2;
+                    var textChars = _lexer.GetSubPattern(beforeBracketPos, _lexer.Position);
+                    var token = CreateToken(RegexKind.TextToken, ImmutableArray<RegexTrivia>.Empty, textChars);
+
+                    // trivia is not allowed anywhere in a character class
+                    ConsumeCurrentToken(allowTrivia: false);
+                    return new RegexPosixPropertyNode(token);
+                }
+                else
+                {
+                    // Reset to back where we were.
+                    // trivia is not allowed anywhere in a character class
+                    _lexer.Position = beforeBracketPos;
+                    ConsumeCurrentToken(allowTrivia: false);
+                    Debug.Assert(_currentToken.Kind == RegexKind.OpenBracketToken);
+                }
+            }
+
+            // trivia is not allowed anywhere in a character class
+            return new RegexTextNode(
+                ConsumeCurrentToken(allowTrivia: false).With(kind: RegexKind.TextToken));
+        }
+
+        private RegexPrimaryExpressionNode ParseCharacterClassSubtractionNode(RegexToken minusToken)
+        {
+            var charClass = ParseCharacterClass();
+
+            if (_currentToken.Kind != RegexKind.CloseBracketToken && _currentToken.Kind != RegexKind.EndOfFile)
+            {
+                minu
```

Comment:
One general concern: how can I tell whether this is complete?
I took a brief look at the .net regex implementation and I see cases that I could not find here: for instance, ["$ pattern"](https://github.com/dotnet/corefx/blob/master/src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs#L1203) (not sure what it is)

How closely does this implementation track to the .nET regex parser?

Reply:
> One general concern: how can I tell whether this is complete?

That is the 64k question :)   
1. you have to trust me.   I spent a lot of time going through their parser line by line, figuring out what it meant and ensuring we would have the same.
2. i went through the entire regex reference here: https://docs.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-language-quick-reference  .  I made sure i parsed everything there and i meticulously tried to ensure we would parse the same as the .net parser.
3. i added several thousands of tests over in the test PR.  The tests not only check our results, but validate htat we parse successfully when .net parses successfully, and we fail (with the same message) when .net fails.

Note: none of this completely verifies that we're 100% with them... but i've come as close as i can with the tools at my disposal.  If you have alternate ways to validate, let me know.

Reply:
> I took a brief look at the .net regex implementation and I see cases that I could not find here: for instance, "$ pattern" (not sure what it is)

the $ pattern is only used in *replacements*.  i.e.  ```Regex.Replace(myStr, "([a-z])([a-z]+)", "$1$2$1")```

I've done no work to understand those strings or have any special IDE features around them.  It's something we can consider later though.

Reply:
> How closely does this implementation track to the .nET regex parser?

My goal is to track exactly.  There is only one known deviation, which i documented in the code here:

```c#
            // Small deviation: Ecmascript allows references only to captures that precede
            // this position (unlike .net which allows references in any direction).  However,
            // because we don't track position, we just consume the entire back-reference.
            //
            // This is addressable if we add position tracking when we locate all the captures.
```

I felt that this deviation was ok, as it was unlikely to be hit in practice, and the cost to make us compliant was very high.  Also, the only downside here is that we do *not* issue an error on a bogus regex  i.e. you'd need a user writing an *incorrect* regex, and then, the only problem we have is we do not report an error.  

Note: it would be very bad for us to report an error on a *good* regex.  That's why there is so much testing that we do not ever do that.